Fix post-merge CI / update ci-jobs util to make base-ref optional

jonathansadowski opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

This PR updates the `ci-jobs` utility such that `baseRef` is no longer a required argument and is now an optional argument. It also removes the `force` option. When no `baseRef` is supplied, `ci-jobs` acts as it did previously when provided the `force` option, returning all jobs for processing. When `baseRef` is supplied, HEAD is compared with baseRef and the changes determine which jobs are returned.

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #43679 .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a branch based on this branch. Make some changes to a file in one of the projects that's included in the CI config within that project's json. For example, add a comment to `packages/js/onboarding/src/utils/countries/index.ts`.
2. Commit those changes to the branch.
3. Switch back to this branch. Run `pnpm utils ci-jobs`. This should return all jobs.
4. Run `pnpm utils ci-jobs --base-ref heads/YOURBRANCHNAME`, replacing `YOURBRANCHNAME` with your branch name. Only the jobs relating to your project should be returned (keeping in mind that changes in dependencies trigger the jobs of their parent). If you updated `packages/js/onboarding/src/utils/countries/index.ts`, this should return jobs from `@woocommerce/onboarding` and `@woocommerce/admin-library` (the latter because onboarding is a dependency of admin-library).
5. Verify that CI passes on this PR and the matrix created by `ci.yml` only runs jobs from monorepo utils. 

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

Commit:
Update ci-jobs util to make base-ref optional

Commit:
Fix linting error

rrennick commented on .github/workflows/ci.yml:
```diff
@@ -38,7 +38,7 @@ jobs:
             }
             const child_process = require( 'node:child_process' );
-            child_process.execSync( 'pnpm utils ci-jobs ' + baseRef );
+            child_process.execSync( `pnpm utils ci-jobs ${ baseRef ? `--base-ref ${ baseRef }` : '' }` );
```
This would be cleaner if line 38 was changed to

```
baseRef  = '--base-ref origin/' + baseRef;
```

Then this line could be

```suggestion
            child_process.execSync( `pnpm utils ci-jobs ${ baseRef  }` );
```

jonathansadowski replied:
Good suggestion. Updated!

rrennick reviewed (commented):
Nice work. I haven't gone through the testing instructions yet. I just had the one comment on code review.

This is probably not related to this PR but the output has some unexpected results which might be due to PRs merged to trunk since this branch was created:

```
# pnpm utils ci-jobs --base-ref trunk
...
Lint Jobs
-  @woocommerce/monorepo-utils - lint
-  @woocommerce/data - lint
-  @woocommerce/product-editor - lint
-  @woocommerce/admin-library - lint
-  @woocommerce/block-library - lint
-  @woocommerce/plugin-woocommerce - lint:php:changes:branch trunk
Test Jobs
-  @woocommerce/monorepo-utils - JavaScript
-  @woocommerce/data - JavaScript
-  @woocommerce/components - JavaScript
-  @woocommerce/product-editor - JavaScript
-  @woocommerce/admin-library - JavaScript
-  @woocommerce/block-library - JavaScript
-  @woocommerce/plugin-woocommerce - PHP
-  @woocommerce/plugin-woocommerce - PHP 8.0
-  @woocommerce/plugin-woocommerce - PHP WP: nightly
-  @woocommerce/plugin-woocommerce - PHP WP: latest - 1
-  @woocommerce/plugin-woocommerce - PHP WP: latest - 2
```

rrennick commented:
> Verify that CI passes on this PR (and the matrix only runs jobs from monorepo utils)

@jonathansadowski There were 32 jobs (10 E2E shards). However, I switched to the latest `trunk and ran `pnpm utils ci-jobs fix/43679-post-merge-ci`. That produced the same list as in my comment above so that isn't an issue with this PR.

rrennick commented:
>Run pnpm utils ci-jobs --base-ref heads/YOURBRANCHNAME, replacing YOURBRANCHNAME with your branch name. Only the jobs relating to your project should be returned (keeping in mind that changes in dependencies trigger the jobs of their parent).

This doesn't produce any results for me. However, I created a branch from `trunk` and committed the same change to `woocommerce.php` and it also produced no results.

The PR looks great to me but we should have testing instructions that produce the expected results.

Commit:
Update baseRef option to be easier to read in ci.yml

jonathansadowski commented:
> There were 32 jobs (10 E2E shards). However, I switched to the latest trunk and ran pnpm utils ci-jobs fix/43679-post-merge-ci`. That produced the same list as in my comment above so that isn't an issue with this PR.

@rrennick thanks for taking a look at this. This PR only affects the jobs from `ci.yml`, the e2e jobs don't come from there. The best way to ensure it's working is to look at the list of jobs produced by that workflow specifically. For example, here's [the run for this PR itself](https://github.com/woocommerce/woocommerce/actions/runs/7646650079?pr=44055). Sorry about the confusion, I've updated the testing instructions to indicate that in the final step.

> This doesn't produce any results for me. However, I created a branch from `trunk` and committed the same change to woocommerce.php and it also produced no results.

Again, my apologies, I should have been more specific re: the types of files that need to be changed. It needs to be a file that matches a pattern from `config.ci` in one (or more) of the project's `package.json` file. I've updated the testing instructions to indicate such, please let me know if they're still unclear or if you run into any issue.

rrennick reviewed (approved):
> I've updated the testing instructions to indicate such, please let me know if they're still unclear or if you run into any issue.

@jonathansadowski Thanks, this tested great.
